### PR TITLE
Created Login, Signup and home page

### PR DESCRIPTION
**Pull request by madhvikokil**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,31 +1,78 @@
...
+  if(auth.uid) {
+    let userType = localStorage.getItem('typeOfUser');
+    if(userType === 'admin') {
+      typeOfUser = 'admin';
+    } else if (userType === 'customer') {
+      typeOfUser = 'customer';
+    } else {
+      typeOfUser = 'seller';
+    }
+  }
```
Comment (chetanbura) [inline src/App.js]: else condition is best for customer

File: src/App.js [hunk 2]
```diff
@@ -1,31 +1,78 @@
...
+
+  const routeComp = routes.map(
+    ({ path, name, componentPath, isExact, authRoute, roles }) => {
+      let Component = React.lazy(() => {
+        return import(`./container/${componentPath}`);
+      });
```
Comment (chetanbura) [inline src/App.js]: move this function outside and use parameter for component path

File: src/App.js [hunk 3]
```diff
@@ -1,31 +1,78 @@
...
-        >
-          Learn React
-        </a>
-        <Divider hidden />
-        <div>
-          <Button content='Primary' primary />
-          <Button content='Secondary' secondary />
-        </div>
-      </header>
+      <Suspense fallback={<p>Loading...</p>}>
```
Comment (chetanbura) [inline src/App.js:62]: Instead using suspense directly, create 1 component for error boundary along with suspense and use that here so you can catch error for that route only

File: src/Hoc/formElement.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import { Input, Radio, Form, TextArea } from 'semantic-ui-react';
+
+const formInput =(props) => {
+  
+    return(
+        <Input
+            type={props.type} 
+            iconPosition='left' 
+            validators={["required"]}
+            errorMessages={["Field Required"]}
```
Comment (chetanbura) [inline src/Hoc/formElement.js]: is every field is required, these values suppose to be from props

File: src/Hoc/formElement.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        <Radio
+            label={props.label}
+            name={props.name}
+            value={props.value}
+            checked={props.checked}
+            onChange={props.onChange}
+        />
+    )
+}
+
```
Comment (chetanbura) [inline src/Hoc/formElement.js]: And try to add common validation logic in this hoc

File: src/component/navMenu.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            name='Sign In'
+            active={activeItem === 'signin'}
+            onClick={this.handleItemClick}
+          />
+          <Menu.Item
+            as={Link} to='/signup'
+            name='Sign Up'
+            active={activeItem === 'signup'}
+            onClick={this.handleItemClick}
+          />
```
Comment (chetanbura) [inline src/component/navMenu.js]: Optimize this with metadata and using map

File: src/container/addBook.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    //       author: props.bookDetail.author,
+    //       status: props.bookDetail.status,
+    //       description: props.bookDetail.description,
+    //       discount: props.bookDetail.discount,
+    //       price: props.bookDetail.price,
+    //       userType: props.bookDetail.userType
+    //     }
+    //   }
+    //   return null
+    // }
```
Comment (chetanbura) [inline src/container/addBook.js]: Remove unwanted code

File: src/container/addBook.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+          const updateDescription = this.state.description !== "" ? this.state.description : this.props.bookDetail.description;
+          const updateAuthor = this.state.author !== "" ? this.state.author : this.props.bookDetail.author;
+          const updateStatus = this.state.status !== "" ? this.state.status : this.props.bookDetail.status;
+          const updatePrice = this.state.price !== "" ? this.state.price : this.props.bookDetail.price;
+          const updateDiscount = this.state.discount !== "" ? this.state.discount : this.props.bookDetail.discount;
+          const updateId = this.state.id !== "" ? this.state.id : this.props.bookDetail.id;
+          const updateUser = this.state.userType !== "" ? this.state.userType : this.props.bookDetail.userType;
+          const updateUid = this.state.uid !== "" ? this.state.uid : this.props.bookDetail.uid;
+          
+          this.setState({ title: updateTitle, description: updateDescription, status: updateStatus, author: updateAuthor, price: updatePrice, id: updateId, discount: updateDiscount, userType: updateUser, uid: updateUid });
```
Comment (chetanbura) [inline src/container/addBook.js]: optimize this

File: src/container/addBook.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+      isAdd !== 'add' && nextProps.bookDetail && this.setState({
+        id: nextProps.bookDetail.id,
+        author: nextProps.bookDetail.author,
+        description: nextProps.bookDetail.description,
+        discount: nextProps.bookDetail.discount,
+        price: nextProps.bookDetail.price,
+        status: nextProps.bookDetail.status,
+        title: nextProps.bookDetail.title,
+        userType: nextProps.bookDetail.userType,
+        uid: nextProps.bookDetail.uid
```
Comment (chetanbura) [inline src/container/addBook.js]: optimize this

File: src/container/addBook.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+                        label: 'Status', options:dropdown, name: "status", placeholder: 'Status', value: status, id: "status", onChange: this.handleSelect,required: true, disabled: isEdit === 'view' ? true : false })}
+                {user === 'admin' && selectElement({
+                        label: 'Seller Type', options:userData, name: "sellerType", placeholder: 'Seller Type', value: `${uid} ${userType}`, id: "sellerType", onChange: this.formChange,required: true, disabled: isEdit === 'view' ? true : false })}
+              </Form.Group>
+              <Form.Group widths='equal'>
+                {formFieldElement({
+                      id: 'form-input-control-price', label: 'Price', placeholder: 'Price', name: 'price', onChange: this.handleChange, value: price, readOnly: isEdit === 'view' ? true : false, required: true, onKeyDown: this.checkNumericNew })} 
+                {formFieldElement({
+                      id: 'form-input-control-discount', label: 'Discount', placeholder: 'Discount', name: 'discount', onChange: this.handleChange, value: discount, readOnly: isEdit === 'view' ? true : false, required: true, onKeyDown: this.checkNumericNew, maxLength: 2 })}
```
Comment (chetanbura) [inline src/container/addBook.js]: optimize this

File: src/container/auth/login.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    
+    if(errorMessages !== []) {
+      this.setState({
+        errorMessages: errorMessages
+      })
+    }
+    if(emailRegexCheck && passwordRegexTest) {
+      this.props.authLogin(this.state);
+      }
+  }
```
Comment (chetanbura) [inline src/container/auth/login.js]: create validation helper where all validations can be handeled

File: src/container/auth/login.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+              <div class="ui stacked segment">
+                <div class="field">
+                    {this.props.formInput({
+                       placeholder: 'E-mail address', type: "text", name: "email", onChange: this.handleChange, value:this.state.email, icon: 'user', error: this.state.emailError })}                
+                </div>
+
+                <div class="field">
+                    {this.props.formInput({
+                       placeholder: 'Password', type: "password", name: "password", onChange: this.handleChange, value:this.state.password, icon: 'lock', error: this.state.passwordError })}                
+                </div>
```
Comment (chetanbura) [inline src/container/auth/login.js]: optimize here and try to handle orChange in hoc only

File: src/container/auth/signup.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+      this.setState({
+        errorMessages: errorMessages
+      })
+    }
+
+    if(emailRegexCheck && passwordRegexTest){
+      this.props.authSignup(this.state);
+      if(this.props.auth.uid) {
+        this.props.history.push("dashboard");
+
```
Comment (chetanbura) [inline src/container/auth/signup.js]: duplicate code

File: src/container/auth/signup.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            </div>
+            <div class="field">
+            {this.props.radioInput({
+                label: 'Customer', name: "typeOfUser", onChange: this.handleRadio, value: 'customer', checked: this.state.typeOfUser === 'customer'})}        
+            {this.props.radioInput({
+                label: 'Seller', name: "typeOfUser", onChange: this.handleRadio, value: 'seller', checked: this.state.typeOfUser === 'seller'})}     
+            </div>
+            <Button
+              color='twitter'
+              disabled={this.state.firstName === '' || this.state.lastName === '' || this.state.email === '' || this.state.password === ''}>Sign Up
```
Comment (chetanbura) [inline src/container/auth/signup.js]: optimize

File: src/container/bookFunctionality.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import { fetchBookDetails } from '../store/action/userAndBookAction';
+class UserList extends React.Component {
+
+    componentDidMount() {
+        this.props.fetchBookDetails();
+    }
+
+    render(){
+        return(
+            <React.Fragment>
```
Comment (chetanbura) [inline src/container/bookFunctionality.js]: `<React.Fragment>` to `<>`   all pages/places

File: src/container/bookList.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+                <Table.Row>
+                    {/* <Table.HeaderCell>Book Id</Table.HeaderCell> */}
+                    <Table.HeaderCell>Title</Table.HeaderCell>
+                    <Table.HeaderCell>Author</Table.HeaderCell>
+                    <Table.HeaderCell>Price(Rs.)</Table.HeaderCell>
+                    <Table.HeaderCell>Discount</Table.HeaderCell>
+                    <Table.HeaderCell>Actual Price</Table.HeaderCell>
+                    <Table.HeaderCell>Status</Table.HeaderCell>
+                    <Table.HeaderCell>User</Table.HeaderCell>
+                    <Table.HeaderCell>Action</Table.HeaderCell>
```
Comment (chetanbura) [inline src/container/bookList.js]: optimize with map and metadata

File: src/container/bookList.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+                    {this.props.bookList && this.props.bookList.map(book => (
+                    <Table.Row key={book.id}>
+                        {/* <Table.Cell>{book.id}</Table.Cell> */}
+                        <Table.Cell>{book.title}</Table.Cell>
+                        <Table.Cell>{book.author}</Table.Cell>
+                        <Table.Cell>{book.price}</Table.Cell>
+                        <Table.Cell>{book.discount}</Table.Cell>
+                        <Table.Cell>{book.actualPrice}</Table.Cell>
+                        <Table.Cell>{book.userType.toUpperCase()}</Table.Cell>
+                        <Table.Cell>{book.status}</Table.Cell>
```
Comment (chetanbura) [inline src/container/bookList.js]: optimize with map and metadata

File: src/container/homePage.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+          {props.publishedBookList && props.publishedBookList.map(book => (
+            <Table.Row key={book.id}>
+            {/* <Table.Cell>{user.id}</Table.Cell> */}
+            <Table.Cell>{book.title}</Table.Cell>
+            <Table.Cell>{book.author}</Table.Cell>
+            <Table.Cell>{book.price}</Table.Cell>
+            <Table.Cell>{book.discount}</Table.Cell>
+            <Table.Cell>{book.actualPrice}</Table.Cell>
+            {/* <Table.Cell>{book.price}</Table.Cell> */}
+            <Table.Cell><Button onClick={() => placeMyOrder(book)}>Place Order</Button></Table.Cell>
```
Comment (chetanbura) [inline src/container/homePage.js]: optimize with map and metadata

File: src/container/myOrders.js [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                    <Table.Body>
+                        {this.props.myOrders && this.props.myOrders.map(order => (
+                        <Table.Row key={order.orderId}>
+                            <Table.Cell>{order.titleOfBook}</Table.Cell>
+                            <Table.Cell>{order.price}</Table.Cell>
+                            <Table.Cell>{order.discount}</Table.Cell>
+                            <Table.Cell>{order.actualPrice}</Table.Cell>
+                            <Table.Cell>{order.user.toUpperCase()}</Table.Cell>
+                            <Table.Cell>{order.status}</Table.Cell>
+                        </Table.Row>
```
Comment (chetanbura) [inline src/container/myOrders.js]: optimize with map and metadata

File: src/container/order.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+                    <Table.Body>
+                        {props.myOrders && props.myOrders.map(order => (
+                        <Table.Row key={order.orderId}>
+                            <Table.Cell>{order.titleOfBook}</Table.Cell>
+                            <Table.Cell>{order.price}</Table.Cell>
+                            <Table.Cell>{order.discount}</Table.Cell>
+                            <Table.Cell>{order.actualPrice}</Table.Cell>
+                            <Table.Cell>{order.status}</Table.Cell>
+                            <Table.Cell>{order.status === 'COMPLETED' ? order.status : 
+                            <Button onClick={() => props.updateOrderByAdmin(order)}>Pending -> Complete</Button>}</Table.Cell>
```
Comment (chetanbura) [inline src/container/order.js]: optimize with map and metadata

File: src/container/userList.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+                    <Table.Body>
+                        {props.userList && props.userList.map(user => (
+                        <Table.Row pointer key={user.email} onClick={() => props.history.push(`/order/${user.uid}`)}>
+                            <Table.Cell>{user.email}</Table.Cell>
+                            <Table.Cell>{user.firstName}</Table.Cell>
+                            <Table.Cell>{user.lastName}</Table.Cell>
+                            <Table.Cell>{user.userType.toUpperCase()}</Table.Cell>
+                        </Table.Row>
+                        ))}
```
Comment (chetanbura) [inline src/container/userList.js]: optimize with map and metadata

**Review by chetanbura — CHANGES_REQUESTED** [review]
Please cover below things
1. Handle Validations in HOC only
2. Optimize code wherever mentions
3. Optimize with metadata for listing records create HOC if possible

File: src/App.js (hunk 1 above)
Comment (madhvikokil) [reply src/App.js]: done

File: src/App.js (hunk 2 above)
Comment (madhvikokil) [reply src/App.js]: done

File: src/App.js (hunk 3 above)
Comment (madhvikokil) [reply src/App.js:62]: done

File: src/Hoc/formElement.js (hunk 4 above)
Comment (madhvikokil) [reply src/Hoc/formElement.js]: done

File: src/Hoc/formElement.js (hunk 5 above)
Comment (madhvikokil) [reply src/Hoc/formElement.js]: done

File: src/component/navMenu.js (hunk 6 above)
Comment (madhvikokil) [reply src/component/navMenu.js]: done

File: src/container/addBook.js (hunk 7 above)
Comment (madhvikokil) [reply src/container/addBook.js]: done

File: src/container/addBook.js (hunk 8 above)
Comment (madhvikokil) [reply src/container/addBook.js]: done

File: src/container/addBook.js (hunk 9 above)
Comment (madhvikokil) [reply src/container/addBook.js]: done

File: src/container/addBook.js (hunk 10 above)
Comment (madhvikokil) [reply src/container/addBook.js]: done

File: src/container/auth/login.js (hunk 11 above)
Comment (madhvikokil) [reply src/container/auth/login.js]: done

File: src/container/auth/login.js (hunk 12 above)
Comment (madhvikokil) [reply src/container/auth/login.js]: done

File: src/container/auth/signup.js (hunk 13 above)
Comment (madhvikokil) [reply src/container/auth/signup.js]: done

File: src/container/auth/signup.js (hunk 14 above)
Comment (madhvikokil) [reply src/container/auth/signup.js]: done

File: src/container/bookList.js (hunk 17 above)
Comment (madhvikokil) [reply src/container/bookList.js]: done

File: src/container/homePage.js (hunk 18 above)
Comment (madhvikokil) [reply src/container/homePage.js]: done

File: src/container/myOrders.js (hunk 19 above)
Comment (madhvikokil) [reply src/container/myOrders.js]: done

File: src/container/order.js (hunk 20 above)
Comment (madhvikokil) [reply src/container/order.js]: done

File: src/container/userList.js (hunk 21 above)
Comment (madhvikokil) [reply src/container/userList.js]: done

File: src/container/bookFunctionality.js (hunk 15 above)
Comment (madhvikokil) [reply src/container/bookFunctionality.js]: done

File: src/container/bookList.js (hunk 16 above)
Comment (madhvikokil) [reply src/container/bookList.js]: done

**Pull request merged**